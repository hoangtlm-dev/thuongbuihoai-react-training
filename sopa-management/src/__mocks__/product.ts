const MOCK_PRODUCTS = [
  {
    name: 'Black shoes',
    price: 87,
    size: ['Small', 'Medium', 'Large'],
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Black',
    image: {
      url: '/images/products/black-shoes.png',
      alt: 'This is black shoes'
    },
    id: '1'
  },
  {
    name: 'Blue shoes',
    price: 45,
    size: ['Small', 'Medium', 'Large'],
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Blue',
    image: {
      url: '/images/products/blue-shoes.png',
      alt: 'This is blue shoes'
    },
    id: '2'
  }
];

const MOCK_PRODUCT = {
  name: 'Black shoes',
  price: 87,
  size: ['Small', 'Medium', 'Large'],
  description:
    'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
  color: 'Black',
  image: {
    url: '/images/products/black-shoes.png',
    alt: 'This is black shoes'
  },
  id: '1'
};

export {
  MOCK_PRODUCTS,
  MOCK_PRODUCT
};
