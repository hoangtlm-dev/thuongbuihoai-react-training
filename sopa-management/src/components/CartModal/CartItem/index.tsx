import {
  useCallback,
  useState
} from 'react';
import toast from 'react-hot-toast';

// Interfaces
import { Product } from '@interfaces/product';

// Components
import Button,
{
  ButtonType
} from '@common/Button';
import Price,
{
  PriceType
} from '@common/Price';
import Text,
{
  ThemeType
} from '@common/Text';
import Icon,
{
  IconType
} from '@components/common/Icon';
import Counter from '@common/Counter';
import Image from '@common/Image';
import PopupDelete from '@common/PopupDelete';

// Stores
import { useCartStore } from '@stores/cart';
import { CONFIRM_MESSAGE } from '@constants/validate';

// Hooks
import { useMutationEditProductInCart } from '@hooks/useMutate';

interface CartItemProps {
  cartItem: Product;
}

const CartItem = ({
  cartItem
}: CartItemProps) => {
  const [openModalConfirm, setOpenModalConfirm] = useState<boolean>(false);
  const { mutate: putProduct } = useMutationEditProductInCart();

  const { deleteCart } = useCartStore();

  const handleIncrement = useCallback(() => {
    putProduct({
      ...cartItem,
      quantity: cartItem.quantity + 1
    });
  }, [cartItem]);

  const handleDecrement = useCallback(() => {
    const checkQuantity = cartItem.quantity > 1
      ? cartItem.quantity - 1
      : cartItem.quantity;

    putProduct({
      ...cartItem,
      quantity: checkQuantity
    });
  }, [cartItem]);

  const handleOpenModalConfirm = useCallback(() => {
    setOpenModalConfirm(!openModalConfirm);
  }, [openModalConfirm]);

  const handleDeleteCart = useCallback(() => {
    deleteCart(cartItem.id);
    toast.success(CONFIRM_MESSAGE.DELETE_SUCCESS);
  }, [cartItem]);

  return (
    <div data-testid='cart-item' className='cart-item'>
      <div className='cart-info'>
        <div className='cart-image'>
          <Image
            className='cart-image-item'
            url={cartItem?.image?.url}
            alt={cartItem.image.alt}
            width={130}
            height={130}
          />
        </div>
        <div className='cart-description'>
          <Text text={cartItem.name} className='cart-text' />
          <Text
            text={`Color: ${cartItem.color}`}
            type={ThemeType.highlightPrimary}
            className='cart-color'
          />
          <Text
            text={`Size: ${cartItem.size}`}
            type={ThemeType.highlightPrimary}
            className='cart-size'
          />
          <div className='card-quantity'>
            <Counter
              value={cartItem?.quantity}
              onDecrement={handleDecrement}
              onIncrement={handleIncrement}
            />
            <Button
              ariaLabel='Trash'
              children={<Icon iconName={IconType.trash} />}
              type={ButtonType.btnIconSecondary}
              className='cart-btn'
              onClick={handleOpenModalConfirm}
            />
            {openModalConfirm && (
              <PopupDelete
                title={CONFIRM_MESSAGE.CONFIRM_DELETE}
                onCancel={handleOpenModalConfirm}
                onDelete={handleDeleteCart}
              />
            )}
          </div>
        </div>
      </div>
      <Price
        value={cartItem.price}
        type={PriceType.tertiary}
        className='cart-price'
      />
    </div>
  );
};

export default CartItem;
