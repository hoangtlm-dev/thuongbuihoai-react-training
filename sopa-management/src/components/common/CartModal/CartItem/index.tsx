import {
  memo,
  useCallback,
  useState
} from 'react';

// Images
import { Trash } from '../../../../../public/images/icons';

// Interfaces
import { Product } from '../../../../interfaces/product';

// Components
import Button,
{
  ButtonType
} from '../../Button';
import Price,
{
  PriceType
} from '../../Price';
import Text,
{
  ThemeType
} from '../../Text';
import Counter from '../../Counter';
import Image from '../../Image';
import PopupDelete from '../../PopupDelete';

// Stores
import { useCartStore } from '../../../../stores/cart';

// Constants
import { VALIDATE_MESSAGE } from '../../../../constants/validate';
import Toast, { ToastType } from '../../../Toast';

interface CartItemProps {
  key?: string;
  cartItem: Product;
}
const CartItem: React.FC<CartItemProps> = ({
  cartItem
}) => {
  const [openModalConfirm, setOpenModalConfirm] = useState<boolean>(false);
  const [toast, setToast] = useState<boolean>(false)

  const {
    increaseQuantity,
    decreaseQuantity,
    deleteCart
  } = useCartStore();

   const handleShowToast = useCallback(() => {
    setToast(!toast);
  }, [toast]);

  const handleIncrement = useCallback(() => {
    increaseQuantity(cartItem.id)
  }, [cartItem]);

  const handleDecrement = useCallback(() => {
    decreaseQuantity(cartItem.id)
  }, [cartItem]);

  const handleOpenModalConfirm = useCallback(() => {
    setOpenModalConfirm(!openModalConfirm);
  }, [openModalConfirm]);

  const handleDeleteCart = useCallback(() => {
    if (!openModalConfirm) {
      handleShowToast();
    }
    deleteCart(cartItem.id);
  }, []);

  return (
    <div data-testid='cart-item' className='cart-item'>
      <div className='cart-info'>
        <figure className='cart-image'>
          <Image className='cart-image-item' url={cartItem?.image?.url} />
        </figure>
        <div className='cart-description'>
          <Text text={cartItem.name} className='cart-text' />
          <Text
            text={`Color: ${cartItem.color}`}
            type={ThemeType.highlightPrimary}
            className='cart-color'
          />
          <Text
            text={`Size: ${cartItem.size}`}
            type={ThemeType.highlightPrimary}
            className='cart-size'
          />
          <div className='card-quantity'>
            <Counter
              value={cartItem?.quantity}
              onDecrement={handleDecrement}
              onIncrement={handleIncrement}
            />
            <Button
              children={<Trash />}
              type={ButtonType.btnIconSecondary}
              className='cart-btn'
              onClick={handleOpenModalConfirm}
            />
            {openModalConfirm && (
              <PopupDelete
                title={VALIDATE_MESSAGE.CONFIRM_DELETE}
                onCancel={handleOpenModalConfirm}
                onDelete={handleDeleteCart}
              />
            )}
            {toast && <Toast title='Success' message='Xoas' type={ToastType.success} />}
          </div>
        </div>
      </div>
      <Price
        value={cartItem.price}
        type={PriceType.tertiary}
        className='cart-price'
      />
    </div>
  );
};

export default memo(CartItem);
