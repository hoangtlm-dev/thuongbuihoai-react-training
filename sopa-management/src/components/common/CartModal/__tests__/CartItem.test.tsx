// Libs
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

// Components
import CartItem from '../CartItem';

// Mocks
import { MOCK_PRODUCT } from '../../../../__mocks__/product';

describe('CartItem component', () => {
  const props = {
    key: '1',
    cartItem: MOCK_PRODUCT,
  };
  test('should render CartItem component', () => {
    const { getByTestId } = render(<CartItem {...props} />);
    expect(getByTestId('cart-item')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<CartItem {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
