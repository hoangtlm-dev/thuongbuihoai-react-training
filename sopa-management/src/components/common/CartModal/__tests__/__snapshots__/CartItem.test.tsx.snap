// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="cart-item"
    data-testid="cart-item"
  >
    <div
      class="cart-info"
    >
      <figure
        class="cart-image"
      >
        <img
          alt=""
          class="cart-image-item"
          data-testid="image"
          src="/images/products/black-shoes.png"
        />
      </figure>
      <div
        class="cart-description"
      >
        <p
          class="cart-text text-"
          data-testid="text"
        >
          Black shoes
        </p>
        <p
          class="cart-color text-highlight-primary"
          data-testid="text"
        >
          Color: Black
        </p>
        <p
          class="cart-size text-highlight-primary"
          data-testid="text"
        >
          Size: Large
        </p>
        <div
          class="card-quantity"
        >
          <div
            class="counter"
            data-testid="counter"
          >
            <button
              class="counter-minus btn btn-icon-primary"
              data-testid="button"
            />
            <div
              class="input-wrapper"
              data-testid="input-value"
            >
              <input
                class="counter-input input-number"
                placeholder=""
                value="3"
              />
            </div>
            <button
              class="counter-plus btn btn-icon-primary"
              data-testid="button"
            />
          </div>
          <button
            class="cart-btn btn btn-icon-secondary"
            data-testid="button"
          >
            <svg
              fill="none"
              height="20"
              viewBox="0 0 20 20"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2.5 5H4.16667H17.5"
                stroke="#02021D"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
              <path
                d="M15.8334 5.00033V16.667C15.8334 17.109 15.6578 17.5329 15.3453 17.8455C15.0327 18.1581 14.6088 18.3337 14.1667 18.3337H5.83341C5.39139 18.3337 4.96746 18.1581 4.6549 17.8455C4.34234 17.5329 4.16675 17.109 4.16675 16.667V5.00033M6.66675 5.00033V3.33366C6.66675 2.89163 6.84234 2.46771 7.1549 2.15515C7.46746 1.84259 7.89139 1.66699 8.33341 1.66699H11.6667C12.1088 1.66699 12.5327 1.84259 12.8453 2.15515C13.1578 2.46771 13.3334 2.89163 13.3334 3.33366V5.00033"
                stroke="#02021D"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
              <path
                d="M8.33325 9.16699V14.167"
                stroke="#02021D"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
              <path
                d="M11.6667 9.16699V14.167"
                stroke="#02021D"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <p
      class="cart-price price-tertiary"
      data-testid="price"
    >
      $87
    </p>
  </div>
</DocumentFragment>
`;
